File: ooc/src/Object.c
#include <stddef.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "Object_private.h"

void * new (const void *_class, ...)
{
    const struct Class* class = _class;
    void* p = calloc(1, class->size);

    assert(p);
    * (const struct Class **) p = class;
    
    if (class->ctor) {
        va_list ap;

        va_start(ap, _class);
        p = class->ctor(p, &ap);
        va_end(ap);
    }

    return p;
}

void delete (void *self)
{
    const struct Class** cp = self;

    if (self && *cp && (*cp)->dtor) {
        self = (* cp)->dtor(self);
    }
    free(self);
}

int differ(const void *self, const void *a)
{
    const struct Class * const * cp = self;
    assert(self && *cp && (* cp)->differ);

    return (* cp)->differ(self, a);
}

size_t sizeOf(const void* self)
{
    const struct Class * const * cp = self;
    assert(self && *cp);

    return (* cp)->size;
}

// int puto(const void *self, FILE *fp);

File: ooc/src/Object.r
#ifndef __OBJECT_PRIVATE_H__
#define __OBJECT_PRIVATE_H__

struct Object {
    const struct Class *class;  /* Object's description */
};

struct Class {
    const struct 		Object _; 				/* class' description */
    const char*         name;					/* class' name */
    const struct Class* super;					/* class' super class */
    size_t              size;					/* class' object's size */
    
    void * (* ctor) (void* self, va_list* app);
    void * (* dtor) (void* self);
    int    (* differ) (const void* self, const void* b);
    int    (* puto) (const void* self, FILE* fp);
};

#endif // __OBJECT_PRIVATE_H__
